reps/2023-06-06-simplify-sync.md

```diff
@@ -0,0 +1,127 @@
+
+These three pain points have grown significantly over the past year, based both on user and internal feedback. This shouldn't be too surprising, as the persistent storage APIs have never been significantly revised before, beyond introduction of the new `storage_path` option, and similarly the Checkpoint API is in beta.
+
+### Should this change be within `ray` or outside?
+
+main `ray` project. Changes are made to Ray Train and Tune.
+
+## Stewardship
+
+### Required Reviewers
+The proposal will be open to the public, but please suggest a few experienced Ray contributors in this technical domain whose comments will help this proposal. Ideally, the list should include Ray committers.
+
+@matthewdeng, @krfricke
+
+### Shepherd of the Proposal (should be a senior committer)
+To make the review process more productive, the owner of each proposal should identify a **shepherd** (should be a senior Ray committer). The shepherd is responsible for working with the owner and making sure the proposal is in good shape (with necessary information) before marking it as ready for broader review.
+
+@pcmoritz
+
+## User Story
+
+Persistence will center around the recently introduced `storage_path` option.
+
+### Persistent trial directory
+
+Ray Train provides the abstraction of a persistent trial directory. Each trial has its current working directory set to a trial specific directory (e.g., `~/ray_results/experiment_1/trial_1`). Train will synchronize this trial directory with persistent storage specified via the `storage_path` option.
+
+This directory is persistent since Train is syncing it to persistent storage for the user. The experiment driver will record experiment results and other metadata to a similar directory on the head node, which will also synchronize to persistent storage.
+
+![overview](user-story.png)
+
+### Checkpoints
+
+While trials can write arbitrary data to the persistent trial directory, they cannot be guaranteed they can read it back on checkpoint/restore. Data that is needed for resume must be recorded via the `train.report(checkpoint)` API.
+
+For example, this is how you could record a Torch checkpoint:
+
+```python
+   torch.save(model, "path/to/dir")
+   train.report(checkpoint=Checkpoint.from_directory("path/to/dir"))
+```
+
+
+Train will take ownership of the specified checkpoint data, deleting the directory after the checkpoint upload finishes. The checkpoint data will then be managed by Train (e.g., Train may restore only the latest checkpoint or delete previous checkpoints to save space according to checkpoint policy).
+
+The files of the recorded checkpoint can be accessed via `result.checkpoint`. The `Checkpoint` object itself is a logical tuple of `(path, pyarrow.fs.FileSystem)`, and this tuple is immutable. Users can also get and set arbitrary metadata to these checkpoints (e.g., preprocessor configs, model settings, etc), which will be recorded in a `metadata.json` file.
```

Comment:
I'm not sure I understand this. The `checkpoint.path` here will always point to the root of the checkpoint dir, and the user shouldn't be able to assume anything about the path above that. This is the only guarantee this API provides.

Reply:
> checkpoint metadata gets loaded back

To be clear, the persisted metadata (metadata.json) shouldn't contain any path data at all.

Reply:
I see, my comment is more about being careful how to do book-keeping on this (path, fs) tuple internally.

For an s3 checkpoint where `storage_path=s3://bucket`, the checkpoint tuple that can be used to download it would be:

`(s3://bucket/exp_dir/trial_dir/ckpt_000, s3 filesystem)`

But, we should make sure not rely on this absolute s3 path for restoration. This is what we currently do to figure out what checkpoint to resume training from. You run into a problem if the s3 bucket gets moved before restoring the run.

Reply:
Right, we shouldn't store absolute path data so that restores can happen even if the data dir was moved.